File: Programmers/Greedy/CCTV.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

struct compare{
	bool operator()(pair<int, int>a, pair<int, int>b){
		return a.second > b.second;
	}
};

int solution(vector<vector<int>> routes) {
    int answer = 0;
    priority_queue<pair<int, int>, vector<pair<int, int>>, compare> pq;
    
    for(int i = 0; i < routes.size(); i++){
        pq.push(make_pair(routes[i][0], routes[i][1]));
    }
    
    while(!pq.empty()){
        int cctv_loc = pq.top().second;
        pq.pop();
        answer++;
        
        while(true){
            if(pq.empty())
                break;
            else if(pq.top().first <= cctv_loc)
                pq.pop();
            else
                break;
        }
    }
    
    return answer;
}

File: Programmers/Greedy/CCTV.py
def solution(routes):
    answer = 0
    length = len(routes)
    routes.sort(key=lambda r : r[1])
    last = 0

    curr = 0
    while curr < length:
        last = routes[curr][1]
        answer += 1

        while True:
            if curr == length:
                break
            elif routes[curr][0] <= last:
                curr += 1
            else:
                break

    return answer

print(solution([[-20,-15],[-14,-5],[-18,-13],[-5,-3]]))